scalyr_agent/copying_manager/worker.py

```diff
@@ -1215,7 +1215,83 @@ def create_shared_object_manager(worker_class, worker_proxy_class):
     """
 
     class _SharedObjectManager(multiprocessing.managers.SyncManager):
-        pass
+        # the PID of the parent process.
+        parent_pid = None
+        # When the shared object manager's process is going to be terminated,
+        # we need to be sure that the worker is stopped first.
+        # This class attribute should provide access to the worker for the watchdog that is implemented below.
+        worker = None
```

Comment:
I think we should pull must of this logic out into a separate WatchDog abstraction that we define in one of the common utlity files.

It's interface should be something like:

```
     class WatchDog():
         def __init__(self, parent_pid, poll_interval, delay_before_kill):
              Here poll_interval is the number of seconds to sleep between checking if
              the parent is still alive.  delay_before_kill is the time to sleep between sending the
              TERM signal and sending the KILL signal.

         def start(self, has_been_stopped_callback):
             Here has_been_stopped_callback is a function that will be invoked by this
             abstraction once the parent is no longer running.  It can be used by the worker to
             make sure it has been stopped.  We might also want to return a bool from this callback
             that indicates if the worker has been stopped because that might change what we log
            out.

Reply:
Good idea. Done.